lib/services/role.go

```diff
@@ -508,31 +929,40 @@ func MatchLabels(selector map[string]string, target map[string]string) bool {
 // for this role set, otherwise it returns ttl unchanges
 func (set RoleSet) AdjustSessionTTL(ttl time.Duration) time.Duration {
 	for _, role := range set {
-		if ttl > role.GetMaxSessionTTL().Duration {
-			ttl = role.GetMaxSessionTTL().Duration
+		maxSessionTTL, err := role.GetOptions().GetDuration(MaxSessionTTL)
+		if err != nil {
+			continue
+		}
+		if ttl > maxSessionTTL.Duration {
+			ttl = maxSessionTTL.Duration
 		}
 	}
 	return ttl
 }
 
-// CheckLogins checks if role set can login up to given duration
-// and returns a combined list of allowed logins
-func (set RoleSet) CheckLogins(ttl time.Duration) ([]string, error) {
+// CheckLoginDuration checks if role set can login up to given duration and
+// returns a combined list of allowed logins.
+func (set RoleSet) CheckLoginDuration(ttl time.Duration) ([]string, error) {
 	logins := make(map[string]bool)
+
 	var matchedTTL bool
 	for _, role := range set {
-		if ttl <= role.GetMaxSessionTTL().Duration && role.GetMaxSessionTTL().Duration != 0 {
+		maxSessionTTL, err := role.GetOptions().GetDuration(MaxSessionTTL)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if ttl <= maxSessionTTL.Duration && maxSessionTTL.Duration != 0 {
 			matchedTTL = true
 		}
-		for _, login := range role.GetLogins() {
+		for _, login := range role.GetLogins(Allow) {
```

Comment:
looking at this logic now, it seems that we should put this `for` loop inside the `if` loop above, because otherwise we are adding logins despite non-matching duration